### PR TITLE
refactor pkg/application to use Appfile

### PR DESCRIPTION
**Pull request by hongchaodeng**

### CODE REVIEW

File: pkg/application/app.go [hunk 1]
```diff
@@ -129,60 +120,6 @@ func (app *Application) Save(envName string) error {
...
-			if NotWorkload(workloadType) {
-				continue
-			}
-			_, ok := workload.(map[string]interface{})
-			if !ok {
-				return fmt.Errorf("format of workload in %s must be map", name)
-			}
-			//TODO(wonderflow) check workload type exists
-			//TODO(wonderflow) check arguments of workload is valid
-		}
```
Comment (wonderflow) [inline pkg/application/app.go:180]: we'd better refactor these logic to appfile instead of remove it.

File: pkg/application/modify.go [hunk 2]
```diff
@@ -2,24 +2,24 @@ package application
...
+	if !ok {
+		s = appfile.Service{}
 	}
-	app.Components[componentName] = map[string]interface{}{
-		workloadType: workloadData,
+	s["type"] = workloadType
+	for k, v := range workloadData {
+		s[k] = v
 	}
-	return app.Validate()
```
Comment (wonderflow) [inline pkg/application/modify.go:22]: why not keep validation here?

File: pkg/application/modify.go [hunk 3]
```diff
@@ -29,51 +29,41 @@ func (app *Application) SetTrait(componentName, traitType string, traitData map[
...
-		return err
+	tm := t.(map[string]interface{})
+	for k, v := range traitData {
+		tm[k] = v
 	}
-	traits[traitType] = traitData
-	comp[Traits] = traits
-	app.Components[componentName] = comp
-	return app.Validate()
+	return nil
```
Comment (wonderflow) [inline pkg/application/modify.go:46]: validation?

File: pkg/application/modify.go [hunk 4]
```diff
@@ -29,51 +29,41 @@ func (app *Application) SetTrait(componentName, traitType string, traitData map[
...
+
+	s, ok := app.Services[componentName]
+	if !ok {
+		return nil
 	}
-	delete(traits, traitType)
-	comp[Traits] = traits
-	app.Components[componentName] = comp
-	return app.Validate()
+	delete(s, traitType)
```
Comment (wonderflow) [inline pkg/application/modify.go:58]: we should put it back into the map, or it won't work

Comment (hongchaodeng) [reply pkg/application/modify.go:58]: Why? I think the map is a pointer and the deletion happens in place

Comment (hongchaodeng) [reply pkg/application/modify.go:58]: See https://play.golang.org/p/PfoAtCmdCrh

**Comment by hongchaodeng** [comment]
Fixed e2e tests and added unit test. Also addressed #413 kubectl issue.

File: pkg/appfile/build.go [hunk 5]
```diff
@@ -22,27 +21,28 @@ type Push struct {
...
+	return b.pushImage(io, image)
 }
 
-func (b *Build) pushImage(io cmdutil.IOStreams) error {
-	io.Infof("pushing image (%s)...\n", b.Image)
+func (b *Build) pushImage(io cmdutil.IOStreams, image string) error {
+	io.Infof("pushing image (%s)...\n", image)
 
 	switch {
 	case b.Push.Local == "kind":
```
Comment (wonderflow) [inline pkg/appfile/build.go:38]: kind不好用，build 到 deploy 的故事不完整。 不过反正已经有docker依赖了，加个 docker push 到Registry的选项？

File: pkg/application/app.go [hunk 6]
```diff
@@ -126,91 +127,58 @@ func (app *Application) Save(envName string) error {
 	if err != nil {
 		return err
 	}
-	return ioutil.WriteFile(filepath.Join(appDir, app.Name+".yaml"), out, 0644)
+	return ioutil.WriteFile(filepath.Join(appDir, app.Name+".yaml"), out, 0o644)
```
Comment (wonderflow) [inline pkg/application/app.go:130]: `0o644 ` 和`0644` 什么区别？

File: pkg/application/app.go [hunk 7]
```diff
@@ -126,91 +127,58 @@ func (app *Application) Save(envName string) error {
...
-		return errors.New("app is nil")
-	}
 	if app.Name == "" {
-		return errors.New("please provide an existed App name")
+		return errors.New("name is required")
 	}
-	if len(app.Components) == 0 {
-		return errors.New("at least one component is required")
+	if len(app.Services) == 0 {
+		return errors.New("at least one service component is required")
```
Comment (wonderflow) [inline pkg/application/app.go:138]: 具体这个名字叫 service还是Component，我们可能得统一一下，不能直接叫 'service component' 吧？

File: pkg/application/app.go [hunk 8]
```diff
@@ -226,177 +194,76 @@ func (app *Application) GetTraitNames(componentName string) ([]string, error) {
...
+	_, config := app.GetServiceConfig(componentName)
+	traitsData := make(map[string]map[string]interface{})
+	for k, v := range config {
+		if !app.tm.IsTrait(k) {
+			continue
 		}
-	case cue.IntKind:
-		switch rv := val.(type) {
-		case float64:
-			return int64(rv)
```
Comment (wonderflow) [inline pkg/application/app.go:289]: 这个有包含在appfile里吗？

File: pkg/commands/up.go [hunk 9]
```diff
@@ -138,21 +145,27 @@ func (o *appfileOptions) ApplyAppConfig(ac *v1alpha2.ApplicationConfiguration) e
...
+	return application.CreateOrUpdateAppConfig(context.TODO(), o.Kubecli, ac)
+}
+
+func (o *appfileOptions) info(name string) {
 	o.IO.Infof("app has been deployed %s%s%s\n", emojiRocket, emojiRocket, emojiRocket)
 	o.IO.Infof("\tURL: http://%s/\n", o.Env.Domain)
-	o.IO.Infof("\tPort forward: vela port-forward %s <port>\n", ac.Name)
-	o.IO.Infof("\tSSH: vela exec %s\n", ac.Name)
-	o.IO.Infof("\tLogging: vela log %s\n", ac.Name)
+	o.IO.Infof("\tPort forward: vela port-forward %s <port>\n", name)
```
Comment (wonderflow) [inline pkg/commands/up.go:167]: vela listen?

File: design/appfile-design.md [hunk 10]
```diff
@@ -112,7 +117,7 @@ spec:
             spec: {
               containers: [{
                 name:  context.name
-                image: context.image
+                image: parameter.image
```
[the remark this answers is not in the document]
Comment (hongchaodeng) [reply design/appfile-design.md:120]: `image` could only be in either parameter or context.

I realize that image needs to be input via CLI as parameters. So I decided to move it out of context.

File: pkg/appfile/build.go (hunk 5 above)
Comment (hongchaodeng) [reply pkg/appfile/build.go:38]: docker push is the default option.

File: pkg/application/app.go (hunk 6 above)
Comment (hongchaodeng) [reply pkg/application/app.go:130]: Actually it is a typo, but output the same number.. fixed

File: pkg/application/app.go (hunk 7 above)
Comment (hongchaodeng) [reply pkg/application/app.go:138]: delete component

File: pkg/application/app.go (hunk 8 above)
Comment (hongchaodeng) [reply pkg/application/app.go:289]: Yes. Otherwise rendering won't succeed

**Review by wonderflow — APPROVED** [review]

**Pull request merged**